WombatProject/Services/BookRepository.cs

```diff
@@ -0,0 +1,129 @@
+﻿using Microsoft.EntityFrameworkCore;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading.Tasks;
+using WombatLibrarianApi.Models;
+
+namespace WombatLibrarianApi.Services
+{
+    public class BookRepository : IBookRepository
+    {
+        protected WombatBooksContext _context { get; }
+
+        public BookRepository(WombatBooksContext context)
+        {
+            _context = context;
+        }
+
+        public async Task<IEnumerable<object>> GetBooksFromBookshelfAsync()
+        {
+            var bookIds = _context.Bookshelves.Select(book => book.BookId).ToList();
+            return await _context.Books
+                .Where(book => bookIds.Contains(book.Id))
+                .Include(bookShelfItem => bookShelfItem.Authors)
+                .Include(bookShelfItem => bookShelfItem.Categories)
+                .Join(_context.Bookshelves,
+                book => book.Id,
+                bookshelf => bookshelf.BookId,
+                (book, bookshelf) => new
+                {
+                    Id = book.Id,
+                    Title = book.Title,
+                    Subtitle = book.Subtitle,
+                    Thumbnail = book.Thumbnail,
+                    Description = book.Description,
+                    PageCount = book.PageCount,
+                    Rating = book.Rating,
+                    RatingCount = book.RatingCount,
+                    Language = book.Language,
+                    MaturityRating = book.MaturityRating,
+                    Published = book.Published,
+                    Publisher = book.Publisher,
+                    BookshelfId = bookshelf.Id
+                })
+                .ToListAsync();
+        }
+
+        public async Task<Bookshelf> GetBookshelfItemByIdAsync(int id)
+        {
+            return await _context.Bookshelves.FindAsync(id);
+        }
+
+        public async Task<Bookshelf> AddBookToBookshelfAsync(Book book)
+        {
+            var bookItem = _context.Books.Where(item => item.Id == book.Id).FirstOrDefault();
+            if (bookItem == null)
+            {
+                _context.Authors.AddRange(book.Authors);
+                _context.Categories.AddRange(book.Categories);
+                _context.Books.Add(book);
+            }
+            var bookshelf = new Bookshelf() { BookId = book.Id };
+            _context.Bookshelves.Add(bookshelf);
+            await _context.SaveChangesAsync();
+            return bookshelf;
+        }
+
+        public async Task<int> RemoveBookFromBookshelfByIdAsync(Bookshelf bookshelf)
+        {
+            _context.Bookshelves.Remove(bookshelf);
+            return await _context.SaveChangesAsync();
+        }
```

Comment:
Suggestion:
```csharp
public async Task RemoveBookFromBookshelfByIdAsync(Bookshelf bookshelf)
        {
            _context.Bookshelves.Remove(bookshelf);
            await _context.SaveChangesAsync();
        }
```

Reply:
Could you explain why removing the the return would be good? The method is called in a controller so I figured that controller would need a return from this one, but I could be mistaken :)

Reply:
The method only does a task (removal), then saves the changes, you don't need any return. Where you called it, you didn't even use its return value, just wanted the method to be executed, so it's nicer this way :)

Reply:
Oh I see! Thanks for explaining it, I will fix that tomorrow!